pkg/trace/api/normalizer_test.go

```diff
@@ -33,277 +34,358 @@ func testSpan() *pb.Span {
 	}
 }
 
+func newTagStats() *info.TagStats {
+	return &info.TagStats{Stats: info.Stats{TracesDropped: &info.TracesDropped{}, TracesMalformed: &info.TracesMalformed{}}}
+}
+
+// statsToIssues returns map of {reason->count} for normalization issues due to dropped or malformed traces
+func statsToIssues(ts *info.TagStats) map[string]int64 {
+	m := ts.TracesDropped.TagValues()
+	for r, c := range ts.TracesMalformed.TagValues() {
```

Comment:
I think this is still not the right approach. The fact that you had to make `TagValues` exported and part of the public API of the `info` package only for it to be used in the tests of another package, is indicative of design problems. The tests we write should not span across testing multiple packages at once and be a hybrid of unit and integration tests. It should either be one or the other.

* `tagValues` converts structure fields to a map. If you want to guarantee that all fields are mapped correctly to the right strings (reasons), that should be done within the `info` package, not here.
* the tests in this package are a mix of an integration and unit test: they don't test that the normalizer correctly update the stats; they test that the normalizer updates the stats _and also_ maps itself to the right key. Do you see the problem here? We should only do the former. The fact this implementation is providing difficulties is a good indication that it is not the right approach.

I would like to kindly ask you to go back to [this suggestion](https://github.com/DataDog/datadog-agent/pull/3738#discussion_r297626197) which is a very clear, contained unit test free of any additional helpers and unnecessary exports. The last part that your tests cover for (that fields map to the right map keys aka. reasons) does not belong in this package, it belongs in the `info` package. The fact the `tagValues` had to become exported only for test purposes is a clear indication of it. There is no reason to complicate things.

If you really wish to write an integration test, you can do so by using the [test package](https://godoc.org/github.com/DataDog/datadog-agent/pkg/trace/test), with some examples found in the [documentation](https://godoc.org/github.com/DataDog/datadog-agent/pkg/trace/test) and real world ones in the [testsuite](https://github.com/DataDog/datadog-agent/tree/master/pkg/trace/test/testsuite) folder. That will allow you to programmatically spin up a trace-agent, a fake backend, send data to it and assert what the backend receives. Then, you can send malformed or bad traces and assert that they don't reach the server or that they do so given the correct transformation.

But please refrain from creating a mix of unit and integration tests. Let's stay purely unit focused. If we test that the normalizer sets the correct stats in this package, and we test that the stats get mapped to the expected keys in the other package, then it is fine to conclude that they will also work together.

If you want to write integrations tests, then please do that as part of the `test` package. I'd say it's quite easy to use and beneficial. I'm happy to help use it if anything is confusing.

Reply:
> The fact that you had to make `TagValues` exported and part of the public API of the `info` package only for it to be used in the tests of another package, is indicative of design problems.

Good point. Ok, updated to compare the structs directly. I don't think we need another integration test on top of this, that may be overkill. This should be ok.

Reply:
I also think it's ok, but I wanted to let you know about it, in case we want to use it. You can even inspect the logs that the agent outputs.